### PR TITLE
Plugin with task for verifying contracts on etherscan

### PR DESCRIPTION
**Pull request by mpetrunic**
Tasks:
- [x] merged boilerplate 
- [x] extended configuration with etherscan prop object
- [x] defined task and params for contract verification
- [x] submit verification request on etherscan
- [x] result polling
- [x] etherscan error handling
- [x] compile contracts and check if contract name exists
- [x] abi encode constructor arguments
- [x] unit and integration tests

this pr should resolve issue #1

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,60 @@
+{
+  "name": "@nomiclabs/buidler-etherscan",
+  "version": "0.0.1",
+  "description": "Buidler TypeScript plugin for verifying contracts on etherscan",
+  "repository": "github:nomiclabs/buidler-ts-plugin-boilerplate",
```
Comment (itirabasso) [inline package.json]: This should be "buidler-etherscan"

File: src/ContractCompiler.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    for (const contract in contracts) {
+      if (contracts.hasOwnProperty(contract) && contract === contractName) {
+        return {
+          abi: contracts[contract].abi,
+          bytecode: contracts[contract].evm.bytecode.object
+        };
+      }
+    }
+    throw new BuidlerPluginError(
+      "Given contract name doesn't exist in sources"
```
Comment (itirabasso) [inline src/ContractCompiler.ts]: I'd show the contract's name in the error.

File: src/etherscan/EtherscanResponse.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+export default class EtherscanResponse {
+  public readonly status: number;
+
+  public readonly message: string;
+
+  public constructor(response: any) {
```
Comment (itirabasso) [inline src/etherscan/EtherscanResponse.ts:6]: It would be nice to have this typed. I couldn't find a _response type_ on request's types tho.

File: src/solc/SolcVersions.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    if (!fullVersion) {
+      throw new BuidlerPluginError("Given solc version doesn't exists");
+    }
+    return fullVersion.replace(/(soljson-)(.*)(.js)/, "$2");
+  }
+
+  public async getVersions(): Promise<{ [key: string]: string }> {
+    try {
+      const response: any = await request.get(
+        "https://raw.githubusercontent.com/ethereum/solc-bin/gh-pages/bin/list.json",
```
Comment (itirabasso) [inline src/solc/SolcVersions.ts]: I'd put this in a constant.

File: src/etherscan/EtherscanVerifyContractRequest.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      throw new BuidlerPluginError(
+        "Failed to parse libraries. Reason: " + e.message
+      );
+    }
+    for (const libraryName in parsedLibraries) {
+      if (parsedLibraries.hasOwnProperty(libraryName)) {
+        this["libraryname" + i] = libraryName;
+        this["libraryaddress" + i] = parsedLibraries[libraryName];
+        i++;
+        if (i >= 10) {
```
Comment (itirabasso) [inline src/etherscan/EtherscanVerifyContractRequest.ts:61]: ~~Why is this needed?~~

Sorry, I just saw it on the etherscan doc.

File: src/index.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+declare module "@nomiclabs/buidler/types" {
+  export interface BuidlerRuntimeEnvironment {
+    etherscan: EtherscanBuidlerEnvironment;
+  }
+
+  export interface ResolvedBuidlerConfig {
+    etherscan: {
+      url?: string;
```
Comment (itirabasso) [inline src/index.ts]: Can this really be undefined? If no url is given the EtherscanBuidlerEnvironment will default to ```https://api.etherscan.io/api ``` on its constructor.

File: src/index.ts [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+      token?: string;
```
Comment (itirabasso) [inline src/index.ts]: I'd call this "api key". Etherscan call it that way and it makes much more sense to me.

File: test/integration/PluginTests.ts [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+  let env: BuidlerRuntimeEnvironment;
+
+  before("Buidler project setup", function() {
+    process.chdir(__dirname + "/../buidler-project");
+    process.env.BUIDLER_NETWORK = "develop";
+    env = require("@nomiclabs/buidler");
+  });
+
+  it("Test verifying deployed contract on etherscan", async () => {
```
Comment (itirabasso) [inline test/integration/PluginTests.ts:28]: I noticed this tests fail because no valid api key is defined. It would be nice to test an actual contract verification (on ropsten), and include it into the CI.

Comment (itirabasso) [reply test/integration/PluginTests.ts:28]: Also, what happen if we reach the 100 daily submissions limit?

File: test/integration/PluginTests.ts [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    const deployedAddress = await deployer.deployContract(
+      compilationResult.abi,
+      "0x" + compilationResult.bytecode,
+      amount
+    );
+    try {
+      await env.run("verify-contract", {
+        address: deployedAddress,
+        contractName: "TestContract1",
+        // libraries: JSON.stringify({
```
Comment (itirabasso) [inline test/integration/PluginTests.ts:56]: Can we add a test case of sources using libraries?

**Review by itirabasso — CHANGES_REQUESTED** [review]
Really nice work! I left some comments, but everything is in good shape.

File: src/etherscan/EtherscanResponse.ts (hunk 3 above)
Comment (alcuadrado) [reply src/etherscan/EtherscanResponse.ts:6]: I'd rather have something like `type EtherscanResponse = any` instead of a direct `any`. This way it'd be easier to change in the future.

File: src/solc/SolcVersions.ts (hunk 4 above)
Comment (alcuadrado) [reply src/solc/SolcVersions.ts]: We have a this typed in Buidler's core. But it's not easy to share that at the moment. Will change it when we finish the migration into a monorepo.

File: src/index.ts [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+    public readonly token: string = ""
+  ) {}
+}
+
+declare module "@nomiclabs/buidler/types" {
+  export interface BuidlerRuntimeEnvironment {
+    etherscan: EtherscanBuidlerEnvironment;
+  }
+
+  export interface ResolvedBuidlerConfig {
```
Comment (itirabasso) [inline src/index.ts]: You should extend BuidlerConfig

something like:
```
export interface EtherscanConfig extends BuidlerConfig {
  etherscan?: {
    url?: string;
    token?: string;
  }
}
```
Notice that etherscan can be undefined too.

File: src/etherscan/EtherscanResponse.ts (hunk 3 above)
Comment (mpetrunic) [reply src/etherscan/EtherscanResponse.ts:6]: Hmm not sure how to proceed on this one. It's common practice to have constructor with any type when casting service response to some model. Not sure what would be the benefit of defining type since every http library will return any type s it ill never be detected at compile time though.

I'm doing some transformation here because status isn't number for some reason, and message prop contains useless OK and NOTOK :-1:

Comment (alcuadrado) [reply src/etherscan/EtherscanResponse.ts:6]: I got your point @mpetrunic. This is fine. Sorry for the misunderstanding.

File: src/etherscan/EtherscanVerifyContractRequest.ts (hunk 5 above)
Comment (mpetrunic) [reply src/etherscan/EtherscanVerifyContractRequest.ts:61]: Yeah, weird as hell. I don't think it's working via API, since it fails on their client as well but works when submitted trough GUI. I'm trying to find where I can report this.

File: test/integration/PluginTests.ts (hunk 9 above)
Comment (mpetrunic) [reply test/integration/PluginTests.ts:56]: It's bugged, as it seems from their sample client. Works trough GUI though, will report it.

**Review by mpetrunic — COMMENTED** [review]
Thanks fr review @itirabasso . I'll address issues in the morning

File: test/integration/PluginTests.ts (hunk 8 above)
Comment (mpetrunic) [reply test/integration/PluginTests.ts:28]: I've changed CI to run only unit tests. These tests are so unstable. Etherscan sometimes needs more than 6-7 confirmations to reach block with deployed contract and this fails with error Contract source not found. Also there is constant need for watching if wallet has funds, and ci will be running for 4-5 mins, which is not ideal. Might be good solution to run it only on master or something.

~~There is no need for valid apikey and there isn't limit on etherscan for testnets. Wallet private key is needed for this and can be suplied via travis env variable~~ suplied it trough travis env

File: src/solc/SolcVersions.ts (hunk 4 above)
Comment (mpetrunic) [reply src/solc/SolcVersions.ts]: I went to use CompilerDownloader but it messes up tests becuase it's caching list.json locally. Would be nice if it provided a flag to prevent caching

File: src/index.ts (hunk 10 above)
Comment (mpetrunic) [reply src/index.ts]: Hmm I'm actually redeclaring `ResolvedBuidlerConfig` to effectively extend it so typescript compiler doesn't complain when I'm using values provided by `buidler.config.js`.

`env.config` [here](https://github.com/nomiclabs/buidler-etherscan/pull/2/files/f1c0b7782d12c2d630263f4082bfe8d4e46cabab#diff-f41e9d04a45c83f3b6f6e630f10117feR39) is of type `ResolvedBuidlerConfig` and without redeclaration it wouldn't know about etherscan props.

Comment (alcuadrado) [reply src/index.ts]: There's been some confusion here. You do need to redeclare a config type, but that's `BuidlerConfig`, not `ResolvedBuidlerConfig`. Note that everything has to be optional, as that's the user-defined config's type.

Something like this should work

```ts
export interface BuidlerConfig {
  etherscan?: {
    url?: string;
    token?: string;
  }
}
```

`ResolvedBuidlerConfig` extends `BuidlerConfig` so you won't have problems with `env.config`.

Comment (mpetrunic) [reply src/index.ts]: makes sense, will update

**Comment by mpetrunic** [comment]
@itirabasso @alcuadrado Is this ok?

File: test/integration/PluginTests.ts (hunk 8 above)
Comment (alcuadrado) [reply test/integration/PluginTests.ts:28]: You are right @mpetrunic, running all the tests in the CI is more complicated than we initially thought.

File: test/integration/PluginTests.ts (hunk 9 above)
Comment (alcuadrado) [reply test/integration/PluginTests.ts:56]: Can you add a comment explaining this? And if you report it, can you link the report in the comment? Thanks!

Comment (mpetrunic) [reply test/integration/PluginTests.ts:56]: They are closed source, so I've reported it trough email. This is what I've sent then so you can try it trough https://ropsten.etherscan.io/sourcecode-demo.html . It fails without explanation. They said that are investigating. 
```
So original deployed source is:
https://gist.github.com/mpetrunic/5f219fd480eb633971db12ef6d17b6d0

Params are  sent as form data but i'm logging it as json are these:
https://gist.github.com/mpetrunic/322e630b332473355ad26b3dec198d86

They work when submitted trough https://ropsten.etherscan.io/verifyContract2?a=0xda628694fece2e6b79ba14cd06435da1c1a95c39 but not when sent to the api, either directly or via https://ropsten.etherscan.io/sourcecode-demo.html.
```

Comment (alcuadrado) [reply test/integration/PluginTests.ts:56]: Thanks for the explanation! Can you get back to us when they answer your email?

Comment (mpetrunic) [reply test/integration/PluginTests.ts:56]: yeah definitely!
